### ISSUE TITLE
PointRend torchscript export with caffe2_traceable 

### ISSUE THREAD

**Issue by bpfliegel** [facebookresearch/detectron2#2397]
## 🚀 Feature
Detectron2 provides a good infrastructure (caffe2_converter.py) to export e.g. MaskRCNN models to Torchscript models. Using standard caffe2_traceable export method models could be successfully exported and inferred.
PointRend mask head traceability was recently implemented in:
https://github.com/facebookresearch/detectron2/commit/b5d90dd2f30aaf4ae567ff3bbcecb1a1ff6e052f
However it seems that PointRend models do not yet have the full support in export scripts (Caffe2XXX patched counterparts of some building blocks) and such models cannot be exported with success.

## Motivation & Examples
PointRend is a fantastic project and allows for excellent output quality. It would be of great benefit if such models could be exported for deployment.

## Note
More of a question: do you have such plans regarding this project? Do you plan to support caffe2_tracing for the PointRend project? Do you also think it would have a high benefit?

Thank you very much,
Bálint

**Comment by ppwwyyxx** [comment]
What's implemented in b5d90dd2f30aaf4ae567ff3bbcecb1a1ff6e052f is for `--export-method tracing`, not `caffe2_tracing`. Support PointRend in caffe2 will be very difficult. 

With `--export-method tracing --format torchscript` and adding pointrend configs, the pointrend models we provided can be exported.

**Comment by bpfliegel** [comment]
Indeed so, but with '--export-method tracing' I could not get a base MaskRCNN model I could successfully infer on Triton r20.11 on GPU (on CPU it works) - it will complain about mixed tensor placement (cpu/cuda). With '--export-method caffe2_tracing' and wrapping tuple inputs (see: https://github.com/triton-inference-server/server/issues/2025) I was able to do so for a base MaskRCNN model for both CPU and GPU. That's why I was insisting on caffe2_tracing - my goal is successful GPU infer with PointRend under Triton.
In order to support caffe2_tracing export method also for PointRend, the following changes should be done:

1) In detectron2/export/c10.py:

Replace
`roi_class_nms = to_device(nms_outputs[2], device)`
with
`roi_class_nms = to_device(nms_outputs[2], device).to(dtype=torch.int64)`
as classes are int64 tensors as per the implementation and several PointRend functions will fail if not so. 

2) In detectron2/export/caffe2_patch.py:

Additionally import:
```
from detectron2.projects.point_rend.roi_heads import PointRendROIHeads
from detectron2.projects.point_rend import mask_head as pointrend_mask_head
```

and replace
```
if getattr(self.heads, "mask_on", False):
    mock_ctx_managers += [mock_mask_rcnn_inference(tensor_mode, mask_head_mod)]
```
with
```
if getattr(self.heads, "mask_on", False):
    if isinstance(self.heads, PointRendROIHeads):
        pointrend_mask_head_mod = pointrend_mask_head.PointRendMaskHead.__module__
        mock_ctx_managers += [mock_mask_rcnn_inference(tensor_mode, pointrend_mask_head_mod)]
    else:      
        mock_ctx_managers += [mock_mask_rcnn_inference(tensor_mode, mask_head_mod)]
```

3) OPTIONAL:
Replace Caffe2MaskRCNNInference class in detectron2/export/c10.py with the following to get N1HW tensors instead of NCHW in order to minimize output data size (and this also corresponds the default implementation).
```
class Caffe2MaskRCNNInference:
    def __call__(self, pred_mask_logits, pred_instances):
        """ equivalent to mask_head.mask_rcnn_inference """
        if all(isinstance(x, InstancesList) for x in pred_instances):
            assert len(pred_instances) == 1
            cls_agnostic_mask = pred_mask_logits.size(1) == 1
            if cls_agnostic_mask:
                mask_probs_pred = pred_mask_logits.sigmoid()
            else:
                num_masks = pred_mask_logits.shape[0]
                class_pred = cat([i.pred_classes for i in pred_instances])
                indices = torch.arange(num_masks, device=class_pred.device)
                mask_probs_pred = pred_mask_logits[indices, class_pred][:, None].sigmoid()
            mask_probs_pred = alias(mask_probs_pred, "mask_fcn_probs")  
            pred_instances[0].pred_masks = mask_probs_pred          
        else:
            mask_rcnn_inference(pred_mask_logits, pred_instances) 
```

After making these changes and wrapping the exported model as per https://github.com/triton-inference-server/server/issues/2025, I am able now to infer PointRend on GPU in Triton r20.11.

@ppwwyyxx  Thanks again for your answer Yuxin, please consider applying my findings.
Thanks, Balint

**Comment by ppwwyyxx** [comment]
`--export-method tracing` should work for GPUs and there are actually unit tests for it.

**Comment by bpfliegel** [comment]
Yes, '--export-method tracing' works fine in a normal Python environment on GPU. But not in the deployment scenario I outlined above - only '--export-method caffe2_tracing' is working.

**Comment by bpfliegel** [comment]
https://github.com/CoderHam should know more about this, will contact him.

**Comment by bpfliegel** [comment]
For any future readers, on second throught: forget this whole attempt above. If you want to infer Detectron2 on GPU in Triton, the easiest and most maintainable method is to use a python backend (with bulding a custom python environment) and implement model.py. Works flawlessly even on multi-GPU. Exports also tend to use more GPU ram - I will prefer this method.

**Comment by ppwwyyxx** [comment]
Just added some example instructions that export a model with tracing and run inference with C++: https://github.com/facebookresearch/detectron2/tree/master/tools/deploy

**Comment by bpfliegel** [comment]
@ppwwyyxx thanks Yuxin for your work, will definitely check that out!
